### PR TITLE
Add C# implemenation for 1D Convolutions

### PR DESCRIPTION
**Pull request by stormofice**
Requires .NET 5.0 [[SDK Install instructions](https://dotnet.microsoft.com/download/dotnet/5.0)]

If you want to test this implementation on your machine, create a file in the `code/csharp` directory called `code.csproj` and paste the following into it:
```
<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net5.0</TargetFramework>
    </PropertyGroup>

</Project>
```
Now you can execute `dotnet run` to run the implementation.

### CODE REVIEW

**Comment by ntindle** [comment]
[lang: c#]

**Comment by Trashtalk217** [comment]
I've attempted to check your code, but I haven't been able to run it yet. C# environments are a whole thing.

I've been able to check the website build, and there's still quite a lot to do there:
image
This is showing up in the Julia implementation and I have no clue why it is there.

If you need help with setting up honkit, please let me know.

**Comment by Trashtalk217** [comment]
I'll do a code review later, since I don't have much time:

The line numbers in the .md file are off for the Julia imports, make sure that looks good.
You've got an off by one error in your .dat files:

image

Here are the files generated by the Julia implementation for reference:
[cyclic.txt](https://github.com/algorithm-archivists/algorithm-archive/files/7107754/cyclic.txt)
[full_linear.txt](https://github.com/algorithm-archivists/algorithm-archive/files/7107755/full_linear.txt)
[simple_linear.txt](https://github.com/algorithm-archivists/algorithm-archive/files/7107756/simple_linear.txt)

Don't forget that Julia starts indexing at 1 for arrays and not 0.

**Comment by stormofice** [comment]
Thanks for the review, I fixed the errors you described above and it now produces the following [seems to be correct] output:
[cyclic.txt](https://github.com/algorithm-archivists/algorithm-archive/files/7107838/cyclic.txt)

**Comment by Trashtalk217** [comment]
The off-by-one error I noted for the cyclic convolution also exists in the other two convolutions.
Also, for the full-linear implementation, Julia generates an output file of 399 lines and you generate 400 lines. I'm guessing that that's a bug in the Julia implementation, however.

**Comment by stormofice** [comment]
Thanks for the review, I forgot about the linear convolutions.

The errors should be fixed now; the full linear output includes a `0.0` at the end which the julia implementation does not, however if you print out the value at `full_linear_output[output_size]` (so the last value in the array) in the julia implementation, it also shows `0.0` and it just does not get written to file (that's my assumption at least).

The following outputs get produced:
[full_linear.txt](https://github.com/algorithm-archivists/algorithm-archive/files/7112529/full_linear.txt)
[simple_linear.txt](https://github.com/algorithm-archivists/algorithm-archive/files/7112530/simple_linear.txt)

**Comment by jiegillet** [comment]
The line missing is probably because we did a ninja edit on the Julia code here: #854. Before that, there was a `0.0` at the end.

**Comment by stormofice** [comment]
Oh it's the same PR that caused the problem with the Julia line numbers, I'll copy those changes then into the C# implementation.

**Comment by stormofice** [comment]
@Trashtalk217 I think that all the aforementioned problems should be fixed now, if you want to take another look at this.

File: contents/convolutions/code/csharp/1DConvolution.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        // Calculates the norm of the array
+        static double Norm(double[] array)
+        {
+            var sum = 0.0;
+            for (var i = 0; i < array.Length; i++)
+                sum += Math.Pow(array[i], 2);
+            return Math.Sqrt(sum);
+        }
+
+        // Julia like modulus function which handles negative values properly
```
Comment (Trashtalk217) [inline contents/convolutions/code/csharp/1DConvolution.cs:34]: If you say what the thing is, you don't have to tell me what it's like.
```suggestion
        // Modulus function which handles negative values properly.
```

File: contents/convolutions/code/csharp/1DConvolution.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        // Assumes that y >= 0
```
Comment (Trashtalk217) [inline contents/convolutions/code/csharp/1DConvolution.cs:35]: The [styleguide](https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions) recommends ending your sentences with a '.' 

```suggestion
        // Assumes that y >= 0.
```

File: contents/convolutions/code/csharp/1DConvolution.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
+﻿using System;
+using System.IO;
+
+namespace Convolution1D
+{
+    public class Convolution1D
+    {
+        // Creates a sawtooth function with the given length
```
Comment (Trashtalk217) [inline contents/convolutions/code/csharp/1DConvolution.cs:8]:
```suggestion
        // Creates a sawtooth function with the given length.
```

File: contents/convolutions/code/csharp/1DConvolution.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        static double[] CreateSawtooth(int length)
+        {
+            var array = new double[length];
+            for (var i = 0; i < length; i++)
+                array[i] = (i + 1) / 200f;
+            return array;
+        }
+
+        // Normalizes the given array
```
Comment (Trashtalk217) [inline contents/convolutions/code/csharp/1DConvolution.cs:17]:
```suggestion
        // Normalizes the given array.
```

File: contents/convolutions/code/csharp/1DConvolution.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        static void Normalize(double[] array)
+        {
+            var norm = Norm(array);
+            for (var i = 0; i < array.Length; i++)
+                array[i] /= norm;
+        }
+
+        // Calculates the norm of the array
```
Comment (Trashtalk217) [inline contents/convolutions/code/csharp/1DConvolution.cs:25]:
```suggestion
        // Calculates the norm of the array.
```

File: contents/convolutions/code/csharp/1DConvolution.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+        // Julia like modulus function which handles negative values properly
+        // Assumes that y >= 0
+        static int Mod(int x, int y) => ((x % y) + y) % y;
+
+        static double[] ConvolveCyclic(double[] signal, double[] filter)
+        {
+            var outputSize = Math.Max(signal.Length, filter.Length);
+
+            // Convolutional output
```
Comment (Trashtalk217) [inline contents/convolutions/code/csharp/1DConvolution.cs:42]:
```suggestion
            // Convolutional output.
```

File: contents/convolutions/code/csharp/1DConvolution.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+                output[i] = sum;
+                sum = 0.0;
+            }
+
+            return output;
+        }
+
+        static double[] ConvolveLinear(double[] signal, double[] filter, int outputSize)
+        {
+            // Convolutional output
```
Comment (Trashtalk217) [inline contents/convolutions/code/csharp/1DConvolution.cs:65]:
```suggestion
            // Convolutional output.
```

File: contents/convolutions/code/csharp/1DConvolution.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                output[i] = sum;
+                sum = 0.0;
+            }
+
+            return output;
+        }
+
+        static void Main()
+        {
+            // Create sawtooth functions for x and y
```
Comment (Trashtalk217) [inline contents/convolutions/code/csharp/1DConvolution.cs:88]:
```suggestion
            // Create sawtooth functions for x and y.
```

File: contents/convolutions/code/csharp/1DConvolution.cs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+            var x = CreateSawtooth(200);
+            var y = CreateSawtooth(200);
+
+            // Normalization is not strictly necessary, but good practice
```
Comment (Trashtalk217) [inline contents/convolutions/code/csharp/1DConvolution.cs:92]:
```suggestion
            // Normalization is not strictly necessary, but good practice.
```

File: contents/convolutions/code/csharp/1DConvolution.cs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+            Normalize(x);
+            Normalize(y);
+
+            // Full convolution, output will be the size of x + y - 1
```
Comment (Trashtalk217) [inline contents/convolutions/code/csharp/1DConvolution.cs:96]:
```suggestion
            // Full convolution, output will be the size of x + y - 1.
```

File: contents/convolutions/code/csharp/1DConvolution.cs [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+            var fullLinearOutput = ConvolveLinear(x, y, x.Length + y.Length - 1);
+            // Simple boundaries
```
Comment (Trashtalk217) [inline contents/convolutions/code/csharp/1DConvolution.cs:98]:
```suggestion
            // Simple boundaries.
```

File: contents/convolutions/code/csharp/1DConvolution.cs [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+            var simpleLinearOutput = ConvolveLinear(x, y, x.Length);
+            // Cyclic convolution
```
Comment (Trashtalk217) [inline contents/convolutions/code/csharp/1DConvolution.cs:100]:
```suggestion
            // Cyclic convolution.
```

File: contents/convolutions/code/csharp/1DConvolution.cs [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+            var cyclicOutput = ConvolveCyclic(x, y);
+
+            // Output convolutions to different files for plotting
```
Comment (Trashtalk217) [inline contents/convolutions/code/csharp/1DConvolution.cs:103]:
```suggestion
            // Output convolutions to different files for plotting.
```

File: contents/convolutions/code/csharp/1DConvolution.cs [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+            File.WriteAllText("full_linear.dat", String.Join(Environment.NewLine, fullLinearOutput));
+            File.WriteAllText("simple_linear.dat", String.Join(Environment.NewLine, simpleLinearOutput));
+            File.WriteAllText("cyclic.dat", String.Join(Environment.NewLine, cyclicOutput));
+        }
+    }
+}
```
Comment (Trashtalk217) [inline contents/convolutions/code/csharp/1DConvolution.cs:109]: You don't have a new line here, and I don't know if that's okay, but Github seems it's important enough to mention it.

**Review by Trashtalk217 — CHANGES_REQUESTED** [review]
Of all things that don't matter, this review matters very little. I checked the build - looks good, and the output - also looks good. There's some minor picking at the comments but it looks great otherwise.

(I also used suggestions so you don't have to go into your editor again, I don't want to waste your time by havng you place periods.)

**Comment by stormofice** [comment]
Should all be fixed now, thanks for the review.

**Review by Trashtalk217 — APPROVED** [review]

**Pull request merged**